src/modules/kvs/cache.c

```diff
@@ -228,44 +223,58 @@ int cache_entry_set_raw (struct cache_entry *hp, void *data, int len)
                 }
             }
         }
-        hp->type = CACHE_DATA_TYPE_RAW;
         return 0;
     }
     return -1;
 }
 
 json_t *cache_entry_get_json (struct cache_entry *hp)
 {
-    if (!hp || !hp->entry_valid || hp->type != CACHE_DATA_TYPE_JSON)
+    if (!hp || !hp->entry_valid || !hp->data)
         return NULL;
-    /* should be non-NULL for json */
-    assert (hp->o);
+    if (!hp->o) {
+        if (!(hp->o = json_loads (hp->data, JSON_DECODE_ANY, NULL)))
+            return NULL;
+    }
     return hp->o;
 }
 
 int cache_entry_set_json (struct cache_entry *hp, json_t *o)
 {
-    if (hp
-        && o
-        && (hp->type == CACHE_DATA_TYPE_NONE
-            || hp->type == CACHE_DATA_TYPE_JSON)) {
+    if (hp && o) {
         if (hp->entry_valid) {
-            assert (hp->o);
-            json_decref (o); /* no-op, 'o' is assumed identical to hp->o */
+            if (hp->data)
+                /* no-op, 'o' is assumed identical to current data */
+                json_decref (o);
+            else if (!hp->data) {
+                /* attempt to change already valid cache entry,
+                 * cannot, must call cache_entry_clear_data() */
+                errno = EBADE;
```

Comment:
this function doesn't set errno in all the -1 return cases (e.g. if args are invalid)

the second test for !hp->data is redundant with the first leg of the conditional.

maybe it would be better to change the hp->entry_valid leg of the outer conditional to simply
```c
if (hp->entry_valid) {
    if (hp->o)
        json_decref (o); // already stored
    else
        hp->o = o;
    assert (hp->len > 0 && hp->data != NULL); // o/w would not be valid json
} else ...
```

Reply:
ugh, why did I add the redundant check :-)

let me think about how to handle the logic here, as I seemed to have copied some logic from the ```set_raw()``` function, but perhaps it shouldn't have been copied over in that way.